Geo/umat parameters python

EleniSmyrniou opened the pull request:
Added UMAT_PARAMETERS as a python variable. This is almost always essential when setting up random field so it is a nice minor improvement

Commit:
Added UMAT_PARAMETERS as python variable

Commit:
Added unit test for the UMAT_PARAMETERS variable

aronnoordam reviewed (changes requested):
placeholder, because I don't remember why umat parameters were required to be exposed to python :D

Commit:
Added functionality to set random field to a vector parameter specifically to "UMAT_PARAMETERS"

Commit:
Typos corrected

Commit:
Weird typo corrected

Commit:
Removed extra definition of start_time that raises an error in the ubuntu build

Commit:
Fixed codacy suggestions

aronnoordam commented on applications/GeoMechanicsApplication/custom_processes/set_parameter_field_process.cpp:
```diff
@@ -28,132 +28,240 @@ namespace Kratos
+        KRATOS_TRY
 
-}  // namespace Kratos.
+            const bool umat_check = mParameters["variable_name"].GetString() == "UMAT_PARAMETERS";
```
as discussed, it might be better to check if the variable is a vector rather than the check if variable == "UMAT_PARAMETERS" but i dont know how, @avdg81 do you have an idea?

aronnoordam replied:
also for now i would change the name of umat_check to: variable_is_umat

aronnoordam commented on applications/GeoMechanicsApplication/custom_processes/set_parameter_field_process.cpp:
```diff
@@ -28,132 +28,240 @@ namespace Kratos
+        KRATOS_CATCH("")
+    }
+
+    void SetParameterFieldProcess::SetParameterFieldForVariableType(const Variable<Vector>& r_var)
```
also as discussed. The code is duplicated, with the difference that r_var has a different type. @avdg81 , can this be shortened?

aronnoordam commented on applications/GeoMechanicsApplication/python_scripts/set_parameter_field_process.py:
```diff
@@ -36,15 +36,41 @@ def __init__(self, Model, settings ):
         self.params.AddValue("dataset", settings["dataset"])
+        is_variable_vector_type = isinstance(self.GetVariableBasedOnString(), KratosMultiphysics.VectorVariable)
+        if "input" in settings["func_type"].GetString() and is_variable_vector_type:
+            self.params.AddValue("vector_index", settings["vector_index"])
```
vector_variable_indices, describes the purpose more I think. Since multiple indices can be given

aronnoordam commented on applications/GeoMechanicsApplication/tests/test_parameter_field.py:
```diff
@@ -12,6 +12,98 @@ class KratosGeoMechanicsParameterFieldTests(KratosUnittest.TestCase):
+        # assert
+        for center_coord, res in zip(center_coords, results):
+            expected_res = 20000 * center_coord[0] + 30000 * center_coord[1]
+            self.assertAlmostEqual(expected_res, res[0][0])
```
can you also check if the non changed values in the umat parameters stay non-changed

aronnoordam commented on applications/GeoMechanicsApplication/custom_processes/set_parameter_field_process.cpp:
```diff
@@ -28,132 +28,240 @@ namespace Kratos
             "dataset"           : "dummy",
-            "dataset_file_name" : "dummy"
+            "dataset_file_name" : "dummy",
+            "vector_index"      : []
```
i think vector_variable_indices is describing the functionality in a better way

aronnoordam commented on applications/GeoMechanicsApplication/custom_processes/set_parameter_field_process.cpp:
```diff
@@ -28,132 +28,240 @@ namespace Kratos
+
+    void SetParameterFieldProcess::SetParameterFieldUsingParametersClass(const Variable<Vector>& rVar, Parameters& rParameters)
+    {
+        const Matrix& r_data_vector = rParameters["values"].GetMatrix();
```
i think you should also use vector_index here, and only set the values at the variable at the index of the vector

aronnoordam replied:
Else I think, the back and fort communication between the external function and thememory usage is unnecessarily inefficient

EleniSmyrniou replied:
Corrected by Anne's changes

aronnoordam commented on applications/GeoMechanicsApplication/tests/test_parameter_field/parameter_field_input_umat_parameters/ProjectParameters.json:
```diff
@@ -0,0 +1,147 @@
+                        "skin_output":         false,
+                        "plane_output":        [],
+                        "nodal_results":       ["DISPLACEMENT","TOTAL_DISPLACEMENT","WATER_PRESSURE","VOLUME_ACCELERATION"],
+                        "gauss_point_results": ["YOUNG_MODULUS","GREEN_LAGRANGE_STRAIN_TENSOR","ENGINEERING_STRAIN_TENSOR","CAUCHY_STRESS_TENSOR","TOTAL_STRESS_TENSOR","VON_MISES_STRESS","FLUID_FLUX_VECTOR","HYDRAULIC_HEAD", "UMAT_PARAMETERS"]
```
can you remove the unused nodal and gauss point results? In order to keep the test small. So i think only "UMAT_PARAMETERS" has to stay

aronnoordam commented on applications/GeoMechanicsApplication/tests/test_parameter_field/parameter_field_json_umat_parameters/ProjectParameters.json:
```diff
@@ -0,0 +1,147 @@
+                        "node_output":         false,
+                        "skin_output":         false,
+                        "plane_output":        [],
+                        "nodal_results":       ["DISPLACEMENT","TOTAL_DISPLACEMENT","WATER_PRESSURE","VOLUME_ACCELERATION"],
```
please remove unused variables

aronnoordam commented on applications/GeoMechanicsApplication/tests/test_parameter_field/parameter_field_python_umat_parameters/ProjectParameters.json:
```diff
@@ -0,0 +1,146 @@
+                        "node_output":         false,
+                        "skin_output":         false,
+                        "plane_output":        [],
+                        "nodal_results":       ["DISPLACEMENT","TOTAL_DISPLACEMENT","WATER_PRESSURE","VOLUME_ACCELERATION"],
```
please remove unused variables

aronnoordam reviewed (commented):
nice, few comments and questions

mcgicjn2 reviewed (changes requested):
Can you make sure that this change doesn't limit the functionality to only UMAT parameters. This should be optional. Checking that UMAT is present means that we limit this change to only those materials, and thus we cannot for instance do this with any of the kratos coded materials parameters (e.g. linear elastic). This should be generic.

avdg81 committed:
Several code improvements

- Removed a few duplicated member functions by introducing member
  function templates.
- The `SetParameterFieldProcess` should now be applicable to any
  `double` variable or `Vector` variable.
- Renamed a few variables.
- Removed a few redundant blank lines and some redundant comments.
- Rewrote a few `for` loops in a more compact way.
- Changed two function parameters to reference-to-const. As a result, we
  can now also pass temporary objects.
- Added a function to get the vector indices. This function also takes
  care of converting the index to the correct type (`double` =>
  `IndexType`).

Commit:
Renamed variable

Commit:
Corrected according to reviewer's comments

Commit:
Merging and renaming values

Commit:
Typing improved

avdg81 commented on applications/GeoMechanicsApplication/custom_processes/set_parameter_field_process.cpp:
```diff
@@ -28,132 +28,132 @@ namespace Kratos
+            "does not have the same size as the amount of elements within the model part!" << std::endl;
+
+        // set new data on the elements
+        const int vector_size = r_data_matrix.size2();
```
Please note that SonarQube reports a narrowing conversion here (i.e. an implicit conversion through which we loose precision). I would suggest you to use the `size_type` of the matrix class rather than narrowing it to an `int`.

EleniSmyrniou replied:
Ok!

Commit:
Type changed to correspond to the matrix

Commit:
Merge branch 'master' into geo/umat_parameters_python

Commit:
Type changed to correspond to the matrix

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -722,6 +722,7 @@ void UPwSmallStrainElement<TDim,TNumNodes>::
     //Defining necessary variables
     const GeometryType& rGeom = this->GetGeometry();
     const IndexType NumGPoints = rGeom.IntegrationPointsNumber( mThisIntegrationMethod );
+    const PropertiesType& rProp = this->GetProperties();
```
The 'this->' is not really necessary here, so unless it is important to re-emphasize it's coming from within the class, I would propose to remove it (it is also personal preference to keep the code a bit cleaner, but I see it's done a lot in this class, so it's up to you if you want to change it or not)

EleniSmyrniou replied:
Indeed maybe we leave this as is because it happens a lot in this class?

rfaasse replied:
That's fine with me

rfaasse reviewed (approved):
The code looks very clean and clear (as a new joiner for the Kratos GeoMechanics team, I learned quite a bit about the UMAT process during this review)! 

I only have one minor suggestion, but that isn't blocking (it's close to a general personal preference, so feel free to implement or ignore it).
